sound/soc/sof/ipc.c

```diff
@@ -251,6 +254,9 @@ int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
 {
 	struct snd_sof_dev *sdev = ipc->sdev;
 	struct snd_sof_ipc_msg *msg;
+	int ret;
+
+	mutex_lock(&ipc->tx_mutex);
 
 	spin_lock_irq(&sdev->ipc_lock);
```

Comment:
Humm, odd. do we need both mutex and spin_lock ?

Reply:
In principle it isn't something extraordinary, but in our case after all these changes it becomes apparent, that we can now also drop the spin-lock. I'll add a patch for that.

Reply:
In fact, no, I think we do need a spin-lock. But not quite the way it is currently used. I will use it to synchronise IPC IRQ threads and the IPC TX context. I think there is currently a race there. Even with the present code we can have:
```
[core 0]
sof_ipc_tx_message()
spin_lock_irq(&sdev->ipc_lock);
list_add_tail(&msg->list, &ipc->tx_list);
spin_unlock_irq(&sdev->ipc_lock);
snd_sof_ipc_msgs_tx(sdev);
<run work-queue>
ipc_tx_next_msg()
spin_lock_irq(&sdev->ipc_lock);
snd_sof_dsp_send_msg(sdev, msg);
                                                      [core 1]
                                                      *** response coming ***
                                                      hda_dsp_ipc_irq_thread()
                                                      reply = snd_sof_ipc_reply(sdev, msg);
                                                      sof_ipc_reply_find_msg()
                                                      *** the above doesn't find the message on the list ***
list_move(&msg->list, &ipc->reply_list);
spin_unlock_irq(&sdev->ipc_lock);
```
So, the receiver will wrongly conclude, that the message isn't a reply. To fix this the sof_ipc_reply_find_msg() must take the same spin-lock as ipc_tx_next_msg(). I'll fix this.

Reply:
I must admit though, that this race is rather theoretical. Core 0 in that code fragment is under spin_lock_irq(), i.e. it has preemption and local IRQs disabled. The chance, that the DSP receives the interrupt, sends a response, it is received by the host, an IRQ thread gets scheduled on core 1, and it checks the reply_list within that race window is probably very close to 0.

Reply:
would be good to add a comment explaining that the use of a spin lock is intentional.